apsuite/optimization/rcds.py

```diff
@@ -431,15 +454,186 @@ def _optimize(self):
-                break
+                break_loop = True
             elif self._num_objective_evals > max_evals:
                 _log.info('Exiting: Maximum number of evaluations reached.')
-                break
+                break_loop = True
 
             func0 = func_min
             pos0 = pos_min
+            self.num_iterations = iter + 1
+            evals_by_iter = len(self.objfuncs_evaluated) - num_evaluations
+            self.num_evals_by_iter.append(evals_by_iter)
             _log.info(
                 f'End of iteration {iter+1:04d}: '
                 f'Final ObjFun = {func_min:.3g}')
+            if break_loop:
+                break
+
+    def plot_history(self):
+        """Plot the history of obj. func. and knobs throughout evaluations.
+
+        Objective function values and knobs values during evaluations are
+        shown in solid lines with transparency, circular markers with no
+        filling signal cumulative optima found during the evaluations and solid
+        filling circular markers refer to RCDS parabolic model inferences at
+        the end of an iteration (guess for the optimum).
+
+        Returns:
+            fig, ax: matplolib figure and axes
+        """
+        opt_idcs, pos_cum_opt, objfuncs_cum_opt = self.get_cumulated_optimum()
+        iters_idcs = _np.concatenate(([0], _np.cumsum(self.num_evals_by_iter)))
+
+        fig, axs = _mplt.subplots(2, 1, figsize=(10, 10), sharex=True)
+        ax = axs[0]
+        ax.plot(
+            self.objfuncs_evaluated,
+            color="C0", alpha=0.4,
+            label="evaluation",
+        )
+        ax.plot(
+            opt_idcs, objfuncs_cum_opt,
+            "o", mfc="none",
+            color="C0",
+            label="cumulated optima",
+        )
+
+        ax.plot(
+            iters_idcs,
+            self.objfuncs_best,
+            "o", color="C0",
+            label="end of iter. optima")
+
+        ax.set_ylabel("objective function")
+        ax.set_xlabel("evaluations")
+
+        ax.legend()
+
+        ax = axs[1]
+        colors = _mplt.rcParams["axes.prop_cycle"].by_key()["color"]
+
+        for i, knob in enumerate(_np.array(self.positions_evaluated).T):
+            color = colors[i % len(colors)]
+
+            ax.plot(knob, alpha=0.4, color=color)
+            ax.plot(
+                opt_idcs,
+                pos_cum_opt[:, i],
+                "o", mfc="none",
+                color=color,
+                label=f"dir {i + 1:2d} - cumulated optima"
+
+            )
+            ax.plot(
+                iters_idcs,
+                self.positions_best[:, i], 'o',
+                color=color,
+                label=f"dir {i + 1:2d} - end of iter. optima"
+                )
+
+        ax.set_ylabel("knobs values")
+        ax.set_xlabel("evaluations")
+        ax.legend()
+
+        fig.tight_layout()
+        fig.show()
+
+        return fig, axs
+
+    def plot_knobspace_slice(self, dir_idcs=(1, 2)):
```

Comment:
couldn't this method be defined at the base class?

Reply:
Same here. Now it is.